_posts/2016-04-07-extension-approach-2.md

```diff
@@ -0,0 +1,85 @@
+logo: 'assets/images/nectar-cart.png'
+---
+>
+The post belongs to _NectarCommerce and Extension Framework Awareness_ Series
+>
+1. _[NectarCommerce Vision](http://vinsol.com/blog/2016/04/08/nectarcommerce-vision/)_
+1. _[Extension Framework Game Plan](http://vinsol.com/blog/2016/04/12/extension-framework-game-plan/)_
+1. _[Introduction to Metaprogramming](http://vinsol.com/blog/2016/04/14/introduction-to-metaprogramming/)_
+1. _[Ecto Model Schema Extension](http://vinsol.com/blog/2016/04/15/ecto-model-schema-extension/)_
+1. Ecto Model Support Functions Extension
+1. Phoenix Router Extension
+1. Phoenix View Extension
+1. Running Multiple Elixir Apps Together
+1. **Extension Approach Explained**
+1. Developer Experience and Workflow developing Favorite Product Extension
+1. Developer Experience and Workflow testing Favorite Product Extension
+
+## What will be NectarCommerce
+
+>
+Off-the-shelf Opensource E-commerce application for building an online store.
+>
+Provides an Extension Framework to support features not included in core as extensions.
+>
+Strives for unobtrusive parallel development of NectarCommerce and Extensions
+
+NectarCommerce is committed to providing a ready-to-use e-commerce solution but the definition of 100% is different under different business domains. It aims to solve common use-cases as part of the project and relying on extension framework to tap the rest.
+
+## NectarCommerce Extension Approach 2 ##
+
+This post aims at documenting how extensions approach 2 is structured, the order of dependencies amongst the various nuts and bolts of NectarCommerce and its extensions as implemented.
+
+NectarCommerce can be divided in two components, both of which reside in the same umbrella application.
+
+1. Nectar: The phoenix project, which acts as the backbone of your E-Commerce application.
+2. Extension Manager: A plugin manager/DSL provider for nectar. It also owns the compile time step for Nectar. We intend to make the DSL/Compiler as a separate package and make Extension Manager solely responsible for downloading and installing extensions.
+
+## Application Overview ##
+
+### Nectar ###
+
+Nectar while being a run of the mill phoenix application, with models, controllers and all has some additional conveniences built into it which makes it amenable to extension. It has no dependencies on other application in the umbrella besides worldly which once done should also be a separate package.
+
+1. __Nectar.Extender__, if any module is using this during compile time, Nectar.Extender searches for whether a module has been compiled which can act as an extension to the module and includes it into the module. There is a basic naming convention followed right now for doing this. If ```Nectar.Product``` uses ```Nectar.Extender``` then it will search for module ```ExtensionsManager.ExtendProduct```. This is a very basic and a proof of concept implementation and we can flesh it out later to support fully name-spaced extension modules.
```

Comment:
Nectar.Extender, if any module is using this during compile time, Nectar.Extender searches for whether a module has been compiled which can act as an extension to the module **(can specify which module)** and includes it into the module.